feat(storage): enable static compaction group

zwang28 opened the pull request:
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Enables e2e static compaction group feature. There're 2 groups for 
  - materialized view and materialized source.
  - internal state and source.
- Assert table ids given to commit_epoch all have been registered to compaction group.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

Little-Wallace commented on src/storage/src/hummock/state_store.rs:
```diff
@@ -81,6 +81,10 @@ impl HummockStorage {
         let epoch = read_options.epoch;
+        let compaction_group_id = match read_options.table_id {
+            None => None,
+            Some(ref table_id) => Some(self.get_compaction_group_id(*table_id).await?),
```
why we need a client to get group id? I think this mapping shall be store in `LocalVersionManager` and be updated with version.

zwang28 replied:
If mapping for a key is not found in client' cache, the whole mapping will be synced.

Little-Wallace commented on src/storage/src/hummock/compaction_group_client.rs:
```diff
@@ -24,27 +24,41 @@ use tokio::sync::RwLock;
+impl CompactionGroupClient for CompactionGroupClientImpl {
     /// Tries to get from local cache
-    pub async fn try_get_compaction_group_id(&self, prefix: Prefix) -> Option<CompactionGroupId> {
+    async fn try_get_compaction_group_id(&self, prefix: Prefix) -> Option<CompactionGroupId> {
```
Do not `async RwLock` because any RPC request will block other read request.

Little-Wallace replied:
suggest `parking_lot::RwLock` to reduce lock wait.

Little-Wallace replied:
```rust
        let mut guard = self.inner.read();
        if let Some(compaction_group_id) = guard.get(&prefix) {
            return Ok(Some(compaction_group_id));
        }
       drop(guard);
        let compaction_groups = self
            .hummock_meta_client
            .get_compaction_groups()
            .await
            .map_err(HummockError::meta_error)?;
                let mut guard = self.inner.write();
        guard.set_index(compaction_groups);
        Ok(guard.get(&prefix))

```

zwang28 replied:
Then we also need to avoid duplicate RPCs.

Little-Wallace replied:
No. This RPC is light and cost little CPU of meta-service. So we do not need to avoid duplicate RPCs

hzxa21 commented on src/storage/src/hummock/compaction_group_client.rs:
```diff
@@ -16,47 +16,56 @@ use std::collections::HashMap;
+#[async_trait::async_trait]
+impl CompactionGroupClient for CompactionGroupClientImpl {
+    /// Tries to get from local cache first,then from meta service
+    async fn get_compaction_group_id(
```
If shared buffer is already split by compaction group, I have a feeling that we will need to know more than just a compaction group id, for example, the compaction group configuration.

Li0k replied:
i think so, hope that `CompactionGroupClient` can fetch more info when rpc, (link configuration) , in order to support ttl and so on

zwang28 replied:
The complete compaction group info is already given to CompactionGroupClient, but CompactionGroupClient only store part of it for now.
I think you can modify it as need later @Li0k .

hzxa21 commented on src/storage/src/hummock/state_store.rs:
```diff
@@ -107,7 +111,7 @@ impl HummockStorage {
         // Generate iterators for versioned ssts by filter out ssts that do not overlap with given
         // `key_range`
-        for level in pinned_version.levels() {
+        for level in pinned_version.levels(compaction_group_id) {
```
In the current implementation, read and write interfaces are bound to a compaction group, which assumes that the mapping between table_id and compaction_group is static and correct. 

I think it is better to leave comments for this assumption since we will need to revisit and refactor all the changes if we want to support dynamic compaction group.

hzxa21 reviewed (approved):
All my comments are concerns on future PRs and not related to "static" compaction group. You can merge this PR first if you want.

zwang28 commented:
@Little-Wallace suggests to update CN's table->group mapping in a more timely manner, rather than current lazy manner.  For example, we can update this mapping  in CN when processing conf change barrier. It makes sense to me.

Also there is an unstable failure parallel e2e test in main. Would only merge this PR after it has been fixed.

zwang28 commented:
> For example, we can update this mapping in CN when CN is process conf change barrier.

I suggest to keep the lazy update approach for now. Because to implement aforementioned more timely approach:
- We need update mapping in LocalStreamManager::send_barrier. It is not async, while update RPC requires async.
- Alternatively it's possible to piggyback the mapping in barrier payload, thus async update RPC is avoided. But I prefer not to do it.

zwang28 commented:
The rpc deduplication implementation is flawed. Will fix before merge.

zwang28 commented:
A summary of changes since last review @Li0k @Little-Wallace @hzxa21 :

- CN's compaction group cache is updated in the beginning of each build_actors.
- Remove prefix concept in compaction group. Use table id directly.
- CompactionGroupClient is quite simple now.

Li0k reviewed (approved):
Thanks for the PR. LGTM!

hzxa21 commented on src/meta/src/stream/stream_manager.rs:
```diff
@@ -531,6 +536,15 @@ where
+
         // In the second stage, each [`WorkerNode`] builds local actors and connect them with
         // channels.
         for (node_id, actors) in node_actors {
```
Do we need to `unregister_table_fragments` if BuildActorsRequest fails?

zwang28 replied:
`purge_stale_members` ensures stale data is eventually cleaned up on each restart.
We can unregister_table_fragments/unregister_source explicitly to clean up more timely. I don't have it here, in order to avoid introducing more error handling code here.

zwang28 replied:
fixed
